sklearn/preprocessing/_polynomial.py

```diff
@@ -923,14 +938,24 @@ def transform(self, X):
         n_splines = self.bsplines_[0].c.shape[1]
         degree = self.degree
 
+        # With sparse=True, we call BSpline.design_matrix which uses extrapolate=False
+        # under the hood.
+        # TODO: With scipy v1.10, use new extrapolate argument in design_matrix(..).
+        # Note: self.bsplines_[0].extrapolate is True for extrapolation in
+        # ["periodic", "continue"]
+        use_sparse = self.sparse and not self.bsplines_[0].extrapolate
+
         # Note that scipy BSpline returns float64 arrays and converts input
         # x=X[:, i] to c-contiguous float64.
         n_out = self.n_features_out_ + n_features * (1 - self.include_bias)
         if X.dtype in FLOAT_DTYPES:
             dtype = X.dtype
         else:
             dtype = np.float64
-        XBS = np.zeros((n_samples, n_out), dtype=dtype, order=self.order)
+        if use_sparse:
+            output_list = []
```

Comment:
There is a bunch of sparse matrices being constructed in this implementation. For each feature:

1. A CSR design matrix is constructed
2. This matrix can be converted into a lil matrix
3. hstack converts it all back to CSR.

What is the runtime for `transform` with sparse output compared to dense output?

Reply:
```
import numpy as np
from sklearn.preprocessing import SplineTransformer

X = np.linspace([-1, -10, 100], [1, 10, 100], 10000)
st_sparse = SplineTransformer(sparse=True, extrapolation="error").fit(X)
st_dense = SplineTransformer(extrapolation="error").fit(X)

%timeit st_dense.transform(X)
2.13 ms ± 20.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit st_sparse.transform(X)
43.7 ms ± 336 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
That's unfortunate.

Reply:
The bottleneck seems to be in scipy's `_bspl.pyx` function `_make_design_matrix`. In contrast, `evaluate_spline` does all loops explicitly. We might report this upstream.

Reply:
@ev-br @egorz734 friendly ping

Reply:
Not sure what's going on here TBH, but csr->lil->csr does sound expensive. Where is the bottleneck, what does the profiler say?

Reply:
My profiler says loud an cleary `_make_design_matrix`.
`%prun st_sparse.transform(X)`gives
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     3    0.046    0.015    0.046    0.015 {scipy.interpolate._bspl._make_design_matrix}
     3    0.003    0.001    0.003    0.001 {built-in method scipy.sparse._sparsetools.coo_tocsr}
    19    0.002    0.000    0.002    0.000 {built-in method builtins.max}
     3    0.002    0.001    0.002    0.001 {built-in method builtins.min}
    26    0.001    0.000    0.001    0.000 {method 'reduce' of 'numpy.ufunc' objects}
     1    0.001    0.001    0.001    0.001 {built-in method scipy.sparse._sparsetools.csr_hstack}
```

Reply:
Given that it's literally a half dozen lines, https://github.com/scipy/scipy/blob/main/scipy/interpolate/_bspl.pyx#L419, I can only suspect the cnp.array buffer protocol. To check this, it looks easy to conver to typed memory views and see if it cuts it. If it does, PR welcome!

Reply:
If not, I'd try to see what yep says (had pleasant experience with it profiling cython code)

Reply:
One other comment is that we (scipy.interpolate) can be open to BSpline.design_matrix growing an optional kwarg for the hstacking, to avoid the csr->lil->csr conversion. 
If this is (roughly) equivalent to numpy ufunc `out=` argument or if it's some other info on the location of a larger csr matrix to place the result to, and if it is something which is generally applicable, not just this specific piece of scikit-learn code, we can consider this. Someone will need to argue for this and provide the implementation though.

Reply:
Most of the LIL conversion will go away with scipy 1.10 and the new `extrapolate` parameter of `design_matrix`. Extending to an ufunc would certainly help, too. And `out` argument would be nice, but preallocating a sparse csr matrix/array might be difficult: We would need to know the non-zero entries in advance.

Reply:
What's the impact of https://github.com/scipy/scipy/pull/16840 on the scikit-learn benchmark in https://github.com/scikit-learn/scikit-learn/pull/24145/files#r940357974?

Reply:
Are there scipy nightly builds to install? Otherwise, this is above my current time budget to benchmark.
Scipy 1.10 will not only speed up `BSpline.design_matrix` but also make it easier to implement spline extrapolation in scikit-learn, thereby further reducing runtime and memory consumption.

Reply:
SciPy has [nightly builds](https://anaconda.org/scipy-wheels-nightly/scipy/files)

```python
pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple scipy
```

but their CI is failing and not updating the nightly builds. I opened https://github.com/MacPython/scipy-wheels/issues/175 to track the SciPy issue.

Reply:
On my laptop, I get (scipy 1.10.1)
```
%timeit st_dense.transform(X)
2.12 ms ± 46.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit st_sparse.transform(X)
4.61 ms ± 162 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```